### ISSUE TITLE
[HOLD for payment 2023-07-20] [$1000] "Incorrect magic code" error if click login in 2FA input page

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21147]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. logout if you already logged in
2. enter email which has 2FA enable
3. enter magic code and click on sign in button
4. scroll login page and click on login
## Expected Result: 
should not show "Incorrect magic code" error
## Actual Result: 
shows "Incorrect magic code" error

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.29-8
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/241fbd82-f742-4537-bfce-c73e3c7d4aa8

https://github.com/Expensify/App/assets/43996225/ce8ad6bf-7435-41f1-b8b5-29c1ae836b15

**Expensify/Expensify Issue URL:**
**Issue reported by:**  @gadhiyamanan 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686808469824509

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010bba1d812d6c5f0f</li>
        <li>Upwork Job ID: 1673472692971782144</li>
        <li>Last Price Increase: 2023-07-03</li>
    </ul>
</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
A sign-in request is unexpectedly fired when pressing the Login button in the Footer while in 2fa form.

### What is the root cause of that problem?
On the Sign In page, it will show the 2FA form only if `showValidateCodeForm` is true.
https://github.com/Expensify/App/blob/1096b6e9cdbdf62a882d29d853a04257d14858fb/src/pages/signin/SignInPage.js#L166
https://github.com/Expensify/App/blob/1096b6e9cdbdf62a882d29d853a04257d14858fb/src/pages/signin/SignInPage.js#L96-L101

One of the requirements to show the form is the `credentials` from Onyx should contain either `login` or `validateCode` property. This means, when the `credentials` is cleared, the form will be hidden.

When we press Login in Footer, it will clear both ACCOUNT and CREDENTIALS onyx
https://github.com/Expensify/App/blob/1096b6e9cdbdf62a882d29d853a04257d14858fb/src/libs/actions/Session/index.js#L569-L574

The problem is when we are clearing both ACCOUNT and CREDENTIALS Onyx, it will notify the update to the subscriber (component) one by one. In our case, SignInPage will first rerender with an empty (null) ACCOUNT. Empty ACCOUNT won't affect the value of `showValidateCodeForm`, it will stay true. This makes the BaseValidateCodeForm a chance to rerender with an empty (null) ACCOUNT.

`BaseValidateCodeForm` is the component for rendering 2FA/magic code form.
https://github.com/Expensify/App/blob/1096b6e9cdbdf62a882d29d853a04257d14858fb/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js#L220-L272

As the ACCOUNT Onyx is cleared, `props.account.requiresTwoFactorAuth` condition becomes false and now it will rerender with magic code instead. When magic code input is mounted, it will trigger the `onFulfill` props because `validateCode` state is never cleared.
https://github.com/Expensify/App/blob/1096b6e9cdbdf62a882d29d853a04257d14858fb/src/components/MagicCodeInput.js#L134-L152

`onFulfill` will trigger the sign in request with the used magic code which then returns an error from the server.

Finally, SignInPage will rerender with an empty CREDENTIALS which hides the form.

<img width="468" alt="image" src="https://github.com/Expensify/App/assets/50919443/8a91b706-7b83-45c8-a138-8b56e75c37b8">

### What changes do you think we should make in order to solve the problem?

Add `&& this.props.account.validated` to `showValidateCodeForm`. This will make sure the form is only shown if `this.props.account.validated` is true and will immediately hide when ACCOUNT is cleared. `this.props.account.validated` will become true once we successfully enter a valid email.

`showValidateCodeForm` is the only condition that doesn't depends on `this.props.account.validated`, so I also think it's good to make it consistent with the others (showPasswordForm, showResendValidationForm, showUnlinkLoginForm)

### What alternative solutions did you explore? (Optional)
Clears CREDENTIALS first by switching up the position of ACCOUNT and CREDENTIALS.
https://github.com/Expensify/App/blob/1096b6e9cdbdf62a882d29d853a04257d14858fb/src/libs/actions/Session/index.js#L569-L574

**Comment by alexpensify** [comment]
This is on my test list

**Comment by alexpensify** [comment]
I'm catching up from my half day, I'll review soon.

**Comment by alexpensify** [comment]
There is a similar GH here: https://github.com/Expensify/App/issues/19712. I'm going to test in Staging to confirm if this one is not fixed by the GH I mentioned.

**Comment by alexpensify** [comment]
The GH above is a different bug but even if it looks similar. I've confirmed that clicking click on Log In button at the button of the page does return this error: 

![image](https://github.com/Expensify/App/assets/8107070/331856d7-e451-4582-86c9-b4bbb8acfe8f)

**Comment by alexpensify** [comment]
@sobitneupane - Out the gate, we have some proposals here. Can you review if any will resolve the issue from this GH? Thank you!

**Comment by alexpensify** [comment]
@sobitneupane - have you been able to review the proposals yet? Thank you for the update!

**Comment by sobitneupane** [comment]
@bernhardoj I kind of like your proposal of handling the issue from `SignInPage`. But we cannot make use of `this.props.account.validated` because I think `this.props.account.validated` is not true for accounts singing in for the first time.

**Comment by bernhardoj** [comment]
@sobitneupane You're right. I didn't think of that case. I updated my proposal to check if the ACCOUNT Onyx is empty or not instead. But with this change, I think I would prefer clearing the CREDENTIALS first solution.

**Comment by alexpensify** [comment]
Reassigning another `bug` team member, I'm going OOO until Monday, July 10, and will take it back then if it's not closed by then. 

Required action from the 🐛 team: Make sure the proposal process is moving forward, so we can get this one assigned out to fix.

**Comment by sobitneupane** [comment]
@bernhardoj I don't think it's a good idea to just swapping the position is a good idea. It might solve this issue but create some unwanted issues like this one which are not easily detected.

https://github.com/Expensify/App/blob/1096b6e9cdbdf62a882d29d853a04257d14858fb/src/libs/actions/Session/index.js#L569-L574


I will prefer addition of some logic. Can you please add more details about `this.props.account.validated` in your proposal? 

> Add && !_.isEmpty(this.props.account.validated) to showValidateCodeForm. This will make sure the form is only shown if ACCOUNT Onyx is not empty and will immediately hide when ACCOUNT is cleared.

**Comment by sobitneupane** [comment]
[Proposal](https://github.com/Expensify/App/issues/21147#issuecomment-1600498246) from @bernhardoj looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by alexpensify** [comment]
Alright, removing the other 🐛 team member. I'm back online and see that this is our current state: 

Prepare for automation!

**Comment by alexpensify** [comment]
Still waiting on automation here

**Comment by alexpensify** [comment]
@gadhiyamanan and @bernhardoj - I've prepared your payments in Upwork. Please accept and I can complete the next steps. 

@sobitneupane please complete the checklist and request payment. 

--- 

Thank you!

**Comment by gadhiyamanan** [comment]
@alexpensify offer accepted, thanks!

**Comment by alexpensify** [comment]
@gadhiyamanan and @bernhardoj - I've completed your payment in Upwork.

**Comment by alexpensify** [comment]
@sobitneupane please complete the checklist and then I can close out this GH. Thank you!

**Comment by alexpensify** [comment]
@anmurali - we are on hold for C+ payment and will give the 🟢 as soon as the checklist has been completed.

**Comment by alexpensify** [comment]
@sobitneupane when you get a chance, please complete the checklist so we can close this one out. Thank you!

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [ ]  [@sobitneupane] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/13655
> * [ ]  [@sobitneupane] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/13655#issuecomment-1648129295
> * [ ]  [@sobitneupane] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

This is an edge case. I don't think this could have been caught in PR reivew.
> * [ ]  [@sobitneupane] Determine if we should create a regression test for this bug.

yes.
> * [ ]  [@sobitneupane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Comment by sobitneupane** [comment]
**Regression Test Proposal**

1. Sign out if you are signed in
2. Enter your email and press Continue
3. Enter the magic code from the email
4. Verify you are brought to the authenticator input page
5. Scroll to the bottom of the Footer
6. Press Login
7. Verify you are scrolled to the top and brought to the login input page
8. Verify there is no error shown below the login input

Do we agree 👍 or 👎

**Comment by alexpensify** [comment]
@sobitneupane, thank you! @jasperhuangg can you please review the checklist and give a 👍🏼 ? Thanks

**Comment by alexpensify** [comment]
Here is the payment summary: 

-  External issue reporter - @gadhiyamanan $250 
-  Contributor that fixed the issue - @bernhardoj $1500
-  Contributor+ that helped on the issue and/or PR - @sobitneupane $1500

Upwork Job: 
https://www.upwork.com/jobs/~010bba1d812d6c5f0f

*_If applicable, the bonuses will be applied on the final payment_

Extra Notes regarding payment: N/A

**Comment by alexpensify** [comment]
@JmillsExpensify - the payment summary has been added here. It's ready for payment-- thank you!

**Comment by JmillsExpensify** [comment]
Reviewed details for @sobitneupane. This is accurate based on summary from Business Reviewer and approved for payment in NewDot.

**Comment by alexpensify** [comment]
Closing - no update here, so I'm going to close this out. I closed the job in Upwork too.
